FirebaseMessaging/Tests/UnitTests/FIRMessagingTokenInfoTest.m

```diff
@@ -0,0 +1,199 @@
+#import "OCMock.h"
+
+static NSString *const kAuthorizedEntity = @"authorizedEntity";
+static NSString *const kScope = @"scope";
+static NSString *const kToken = @"eMP633ZkDYA:APA91bGfnlnbinRVE7nUwJSr_k6cuSTKectOlt66dKv1r_-"
+                                @"9Qvhy9XljAI62QPw307rgA0MaFHPnrU5sFxGZvsncRnkfuciwTUeyRpPNDZMFhNXt"
+                                @"7h1BKq9Wb2A0LAANpQefrPHVUp4p";
+static NSString *const kFirebaseAppID = @"firebaseAppID";
+static NSString *const kIID = @"eMP633ZkDYA";
+static BOOL const kAPNSSandbox = NO;
+
+@interface FIRMessagingTokenInfoTest : XCTestCase
+
+@property(nonatomic, strong) NSData *APNSDeviceToken;
+@property(nonatomic, strong) FIRMessagingTokenInfo *validTokenInfo;
+@property(nonatomic, strong) id mockOptions;
+
+@end
+
+@implementation FIRMessagingTokenInfoTest
+
+- (void)setUp {
+  [super setUp];
+
+  self.APNSDeviceToken = [@"validDeviceToken" dataUsingEncoding:NSUTF8StringEncoding];
+
+  self.mockOptions = OCMClassMock([FIROptions class]);
+  OCMStub([self.mockOptions defaultOptionsDictionary]).andReturn(@{
+    kFIRGoogleAppID : kFirebaseAppID
+  });
+
+  self.validTokenInfo =
+      [[FIRMessagingTokenInfo alloc] initWithAuthorizedEntity:kAuthorizedEntity
+                                                        scope:kScope
+                                                        token:kToken
+                                                   appVersion:FIRMessagingCurrentAppVersion()
+                                                firebaseAppID:FIRMessagingFirebaseAppID()];
+  self.validTokenInfo.APNSInfo =
+      [[FIRMessagingAPNSInfo alloc] initWithDeviceToken:self.APNSDeviceToken
+                                              isSandbox:kAPNSSandbox];
+  self.validTokenInfo.cacheTime = [NSDate date];
+
+  [[NSUserDefaults standardUserDefaults] setObject:FIRMessagingCurrentLocale()
+                                            forKey:kFIRMessagingInstanceIDUserDefaultsKeyLocale];
+}
+
+- (void)tearDown {
+  [self.mockOptions stopMocking];
+  [super tearDown];
+}
+
+- (void)testTokenInfoCreationWithInvalidArchive {
+  NSData *badData = [@"badData" dataUsingEncoding:NSUTF8StringEncoding];
+  FIRMessagingTokenInfo *info = nil;
+  @try {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    info = [NSKeyedUnarchiver unarchiveObjectWithData:badData];
+#pragma clang diagnostic pop
+  } @catch (NSException *e) {
+    XCTAssertEqualObjects([e name], @"NSInvalidArgumentException");
+  }
+  XCTAssertNil(info);
+}
+
+// Test that archiving a FIRMessagingTokenInfo object and restoring it from the archive
+// yields the same values for all the fields.
+- (void)testTokenInfoEncodingAndDecoding {
+  FIRMessagingTokenInfo *info = self.validTokenInfo;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  NSData *archive = [NSKeyedArchiver archivedDataWithRootObject:info];
```

Comment:
optional: I would suggest using [GULSecureCoding](https://github.com/google/GoogleUtilities/blob/1e0f385577a122bd935881cdcb24dbd1aa642663/GoogleUtilities/Environment/Public/GoogleUtilities/GULSecureCoding.h#L28) to avoid using deprecated API when unnecessary (here and at similar places).

Reply:
that's on my todo list next because we did introduce the NSCoding from IID to Messaging that we should migrate to NSSecureCoding.